### PR TITLE
Add Image Transfer resources

### PR DESCRIPTION
**Pull request by rgolangh**
Motivation
OKD on oVirt is using terraform to stand up a cluster, by
creating masters, all from a specific qcow url. This url changes with the
version and needs to uploaded per the cluster version. The intention is to
make it as seemless as possible for users that install OKD on oVirt,
without a pre-step of creating that specific template.

Modification
Add an ovirt_image_resource, that will handle disk uploads
from urls, or local files (which counts as urls). oVirt api exposes 1 or 2
valid urls to upload the image to: either a proxy url, when the proxy is
the only accesible node for the client, and the host url. The provider will
first try to reach the host_url, which is quicker, and if its not available
it will use the proxy url. Update on and image transfer is not supported.

Result
Given this main.tf file, a disk with the name rhcos will be created 
under the provided storage domain id:

```tf
resource "ovirt_image_transfer" "rhcos_transfer" {
  alias             = "rhcos"
  source_url        = "https://path/to/image"
  storage_domain_id = "d787bf6b-fae1-4a3e-b773-2ac466599d29"
  sparse            = true
}
```

### CODE REVIEW

File: ovirt/resource_ovirt_image_transfer.go [hunk 1]
```diff
@@ -0,0 +1,367 @@
...
+		log.Printf("failed to parse the proxy url (%s) : %s", transfer.MustProxyUrl(), err)
+		return "", err
+	}
+	optionsReq, err := http.NewRequest(http.MethodOptions, proxyUrl.String(), strings.NewReader(""))
+	res, err := insecureClient.Do(optionsReq)
+	log.Printf("OPTIONS call on %s: %v", proxyUrl.String(), res.StatusCode)
+	if err == nil && res.StatusCode == 200 {
+		return proxyUrl.String(), nil
+	}
+	return "", err
```
Comment (rgolangh) [inline ovirt/resource_ovirt_image_transfer.go]: actually I fail it here. Why did I return a "" ?

File: ovirt/resource_ovirt_image_transfer.go [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+				// should occur only if the transfer was deleted
+				return nil, "", nil
+			}
+			return nil, "", err
+		}
+
+		return r.MustImageTransfer(), string(r.MustImageTransfer().MustPhase()), nil
+	}
+}
+
```
Comment (rgolangh) [inline ovirt/resource_ovirt_image_transfer.go]: I'll add a reference for that header parsing

**Comment by rgolangh** [comment]
@imjoey PTAL

File: ovirt/resource_ovirt_image_transfer.go [hunk 3]
```diff
@@ -0,0 +1,368 @@
...
+		return err
+	}
+	// already read 32 bits. Look for an abstraction to would allow
+	// seeking back.
+	sFile.Close()
+	// open again
+	sFile, err = os.Open(sFile.Name())
+	if err != nil {
+		return err
+	}
```
[the remark this answers is not in the document]
Comment (imjoey) [reply ovirt/resource_ovirt_image_transfer.go]: @rgolangh  sorry for the late response. I thought I've leave a comment several days ago but it seems to be not committed and still in pending state.

[the remark this answers is not in the document]
Comment (rgolangh) [reply ovirt/resource_ovirt_image_transfer.go]: Wonderful! thanks for mentioning the Seek, don't know how I missed it!

**Comment by rgolangh** [comment]
I'm dumpig the image_transfer representation and moving this funcionality under 'disk' - image transfer is just a transient object created in order to eventually create a disk resource, from url.
So essentially I want to enable another way to create the disk resource, from a source_url.

@imjoey thoughts on this?

**Comment by imjoey** [comment]
@rgolangh  yes, that's true. Unlike ansible, terraform can not manage transient objects. If we add an extra `source_url` attribute for resource `ovirt_disk`, can we also retrieve the value of `source_url` via REST API anywhere? Thanks

**Comment by rgolangh** [comment]
> @rgolangh yes, that's true. Unlike ansible, terraform can not manage transient objects. If adding an extra `source_url` attribute for resource `ovirt_disk`, it would be better that the value of `source_url` can be retrieved via REST API. If not, we may need to move it out of the control of terraform. Thanks.

In the end I stayed with an image_transfer resource. Seems like a disk from source needs more changes than I thought initially and this may end up more intrusive change because when transfering or importing a disk we don't have all the mandatory parameters to create the disk. This means that the scheme would have to be conditional - i.e if source_url is defined we don't have size and format yet, it depends on imported disk.

**Comment by imjoey** [comment]
> In the end I stayed with an image_transfer resource. Seems like a disk from source needs more changes than I thought initially and this may end up more intrusive change because when transfering or importing a disk we don't have all the mandatory parameters to create the disk. This means that the scheme would have to be conditional - i.e if source_url is defined we don't have size and format yet, it depends on imported disk.

@rgolangh totally agreed. An independent image_transfer would keep the disk importing simple. Thanks for updating.

**Comment by rgolangh** [comment]
Can we merge this PR?

**Comment by imjoey** [comment]
@rgolangh  Sure, thank you for this great work.  Everything looks good to me, while could you please format the example `.tf` files before merging? Thanks.

**Comment by rgolangh** [comment]
done

**Comment by imjoey** [comment]
@rgolangh Thank you very much. 😄

File: ovirt/resource_ovirt_image_transfer.go (hunk 2 above)
Comment (giulianoquites) [reply ovirt/resource_ovirt_image_transfer.go]: Hi rgolangh,

To import the image with terraform. The images are pending. When this happens, it is usually due to the certificate.

I did the test by passing the CA certificate and without passing the certificate. Either way, the disk is pending to be imported into the data domain.

Is there any way to test this problem?

Thanks.

**Pull request merged**